File: src/main/java/org/starlisp/core/LispFlonum.java
package org.starlisp.core;

import java.math.BigInteger;

public final class LispFlonum extends LispNumber {
  private double n;

  public static LispNumber parse(String str) {
    return new LispFlonum(Double.parseDouble(str));
  }

  public LispFlonum(double nbr) {
    n = nbr;
  }

  public int hashCode() {
    return Double.valueOf(n).hashCode();
  }

  public LispFlonum add(LispFlonum nbr) {
    return new LispFlonum(n + nbr.n);
  }

  public LispFlonum sub(LispFlonum nbr) {
    return new LispFlonum(n - nbr.n);
  }

  public LispFlonum mul(LispFlonum nbr) {
    return new LispFlonum(n * nbr.n);
  }

  public LispFlonum div(LispFlonum nbr) {
    return new LispFlonum(n / nbr.n);
  }

  public boolean negP() {
    return n < 0;
  }

  public boolean equals(Object obj) {
    return (obj instanceof LispFlonum) ? n == ((LispFlonum) obj).n :
        (obj instanceof LispInteger) ? n == ((LispInteger) obj).toJavaDouble() :
            false;
  }

  public LispFlonum add(LispNumber nbr) {
    return (nbr instanceof LispInteger) ? add(coerceIntegerToFlonum(nbr)) : add((LispFlonum) nbr);
  }

  public LispFlonum sub(LispNumber nbr) {
    return (nbr instanceof LispInteger) ? sub(coerceIntegerToFlonum(nbr)) : sub((LispFlonum) nbr);
  }

  public LispFlonum mul(LispNumber nbr) {
    return (nbr instanceof LispInteger) ? mul(coerceIntegerToFlonum(nbr)) : mul((LispFlonum) nbr);
  }

  public LispFlonum div(LispNumber nbr) {
    return (nbr instanceof LispInteger) ? div(coerceIntegerToFlonum(nbr)) : div((LispFlonum) nbr);
  }

  public String toString() {
    return "" + n;
  }

  public int toJavaInt() {
    return (int) n;
  }

  public long toJavaLong() {
    return (long) n;
  }

  public float toJavaFloat() {
    return (float) n;
  }

  public double toJavaDouble() {
    return (double) n;
  }

  public BigInteger toJavaBigInteger() {
    return BigInteger.valueOf((long) n);
  }
}

File: src/main/scala/org/starlisp/core/LispFlonum.scala
package org.starlisp.core

import java.math.BigInteger

object LispFlonum {
  def parse(str: String): LispNumber = new LispFlonum(str.toDouble)
}

final class LispFlonum(val n: Double) extends LispNumber {

  override def hashCode: Int = n.hashCode

  def add(nbr: LispFlonum): LispFlonum = new LispFlonum(n + nbr.n)
  def sub(nbr: LispFlonum): LispFlonum = new LispFlonum(n - nbr.n)
  def mul(nbr: LispFlonum): LispFlonum = new LispFlonum(n * nbr.n)
  def div(nbr: LispFlonum): LispFlonum = new LispFlonum(n / nbr.n)
  def negP: Boolean = n < 0

  override def equals(obj: Any): Boolean = {
    if ((obj.isInstanceOf[LispFlonum])) n == (obj.asInstanceOf[LispFlonum]).n else if ((obj.isInstanceOf[LispInteger])) n == (obj.asInstanceOf[LispInteger]).toJavaDouble else false
  }

  def add(nbr: LispNumber): LispFlonum = {
    if ((nbr.isInstanceOf[LispInteger])) add(coerceIntegerToFlonum(nbr)) else add(nbr.asInstanceOf[LispFlonum])
  }

  def sub(nbr: LispNumber): LispFlonum = {
    if ((nbr.isInstanceOf[LispInteger])) sub(coerceIntegerToFlonum(nbr)) else sub(nbr.asInstanceOf[LispFlonum])
  }

  def mul(nbr: LispNumber): LispFlonum = {
    if ((nbr.isInstanceOf[LispInteger])) mul(coerceIntegerToFlonum(nbr)) else mul(nbr.asInstanceOf[LispFlonum])
  }

  def div(nbr: LispNumber): LispFlonum = {
    if ((nbr.isInstanceOf[LispInteger])) div(coerceIntegerToFlonum(nbr)) else div(nbr.asInstanceOf[LispFlonum])
  }

  override def toString: String = String.valueOf(n)

  def toJavaInt: Int = n.asInstanceOf[Int]
  def toJavaLong: Long = n.asInstanceOf[Long]
  def toJavaFloat: Float = n.asInstanceOf[Float]
  def toJavaDouble: Double = n
  def toJavaBigInteger: BigInteger = BigInteger.valueOf(n.asInstanceOf[Long])
}